R/__ddl_by_AC.R

```diff
@@ -0,0 +1,298 @@
+    join_keys = join_keys
+  )
+  class(ans) <- c("ddl", class(ans))
+  ans
+}
+
+
+#' wrapper for `on_submit` functions
+#'
+#' Wrap a function that makes some assignments in its body to return a `tdata` object with optionally masked code.
+#'
+#' Code found in the body of `fun` will be run in order to obtain the desired data sets.
+#' References to `input$<name>` will be substituted with input values of the accompanying `shiny` module
+#' for the purposes of code execution. If `mask` is provided, those references will be substituted with mask values
+#' for the purposes of storing code.
+#'
+#' @param fun a function that takes exactly one argument, `input`, which is a named list
+#' @param mask optional named list to specify code masking; see `input_template` for details
+#' @param join_keys optional `join_keys` object; see `input_template` for details
+#'
+#' @return
+#' A `tdata` object containing variables that were created in the body of `fun`
+#' and the entirety of the body of `fun` in the `@code` slot.
+#'
+#' @keywords internal
+with_substitution <- function(fun, mask, join_keys) {
+  checkmate::assert_true(
+    identical(names(formals(fun)), "input"),
+    .var.name = "'on_submit' function only takes 'input' argument"
+  )
+  checkmate::assert_list(mask, names = "unique")
+  checkmate::assert_r6(join_keys, "JoinKeys")
+
+  function(...) {
+    # Get input values from call arguments.
+    call_args <- as.list(match.call(fun))$input
+    checkmate::assert_list(call_args, names = "unique", .var.name = "input")
+    # Add non-masked arguments to mask.
+    mask <- c(mask, call_args)
+    mask <- mask[!duplicated(names(mask))]
+
+    # Extract function body as list of calls.
+    fun_body <- body(fun)
+    code <-
+      if (is.expression(fun_body)) {
+        as.list(fun_body)
+      } else if (is.call(fun_body)) {
+        if (identical(as.list(fun_body)[[1L]], as.symbol("{"))) {
+          as.list(fun_body)[-1L]
+        } else {
+          list(fun_body)
+        }
+      } else if (is.name(fun_body)) {
+        fun_body
+      } else {
+        stop("with_substitution: don't know ho to handle this kind of function body")
+      }
+
+    # Convert calls to strings and substitute argument references by bquote references.
+    code_strings <- vapply(code, deparse1, character(1L))
+    code_strings <- gsub("(input\\$)(\\w+)", "\\.(\\2\\)", code_strings)
+    code_strings <- gsub("(input\\[\\[\")(\\w+)(\"\\]\\])", "\\.(\\2\\)", code_strings)
+    # Use bquote to obtain code with input values and masking values.
+    code_input <- lapply(code_strings, function(x) do.call(bquote, list(str2lang(x), call_args)))
+    code_masked <- lapply(code_strings, function(x) do.call(bquote, list(str2lang(x), mask)))
+
+    # Evaluate input code in separate environment.
+    env <- new.env()
+    eval(as.expression(code_input), env)
+    # Create `tdata` with masked code.
+    new_tdata(as.list(env), code = as.expression(code_masked), keys = join_keys)
```

Comment:
We need a way to change `code_masked` when creating `tdata` object

```suggestion
    new_tdata(as.list(env), code = as.expression(code_masked), keys = join_keys)
```


There is a case where in the interactive session one can use `login()` function which opens a dialog, while for shiny session authentication is handled by the module and no `login()` function is used. It means that initial `code_masked` looks like this

```r
data <- connection_package::read_data(...)
```

and returned should look like:

```r
connection_package::login()
data <- connection_package::read_data(...)
```